packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts

```diff
@@ -379,6 +410,8 @@ export const createColumnsState = ({
     });
   }
 
+  columnsState = updatePinnedColumns(columnsState as GridColumnsState, theme);
```

Comment:
Pinning the checkbox selection column acts weird:

<img width="424" src="https://github.com/mui/mui-x/assets/13808724/29614088-3c9e-4384-8c9b-ba23360164cb">

Here's a minimal repro (couldn't link the codesandbox because CSCI has failed):

```tsx
import * as React from 'react';
import { DataGridPro, GRID_CHECKBOX_SELECTION_FIELD } from '@mui/x-data-grid-pro';
import { useDemoData } from '@mui/x-data-grid-generator';

export default function Demo() {
  const { data } = useDemoData({
    dataSet: 'Commodity',
    rowLength: 100,
    maxColumns: 20,
  });

  return (
    <div style={{ height: 600 }}>
      <DataGridPro
        {...data}
        checkboxSelection
        initialState={{
          pinnedColumns: { left: [GRID_CHECKBOX_SELECTION_FIELD, 'desk'] },
        }}
      />
    </div>
  );
}
```

I looked into it and this diff solves the issue:

```diff
diff --git a/packages/grid/x-data-grid-pro/src/hooks/features/columnPinning/useGridColumnPinningPreProcessors.ts b/packages/grid/x-data-grid-pro/src/hooks/features/columnPinning/useGridColumnPinningPreProcessors.ts
index 7c92b7890..1ab69aaea 100644
--- a/packages/grid/x-data-grid-pro/src/hooks/features/columnPinning/useGridColumnPinningPreProcessors.ts
+++ b/packages/grid/x-data-grid-pro/src/hooks/features/columnPinning/useGridColumnPinningPreProcessors.ts
@@ -5,7 +5,11 @@ import {
   gridPinnedColumnsSelector,
   useGridRegisterPipeProcessor,
   eslintUseValue,
+  updatePinnedColumns,
+  gridVisibleColumnFieldsSelector,
 } from '@mui/x-data-grid/internals';
+import { useTheme } from '@mui/material/styles';
+import { GridColumnsState } from '@mui/x-data-grid';
 import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
 import { GridPrivateApiPro } from '../../../models/gridApiPro';
 
@@ -24,10 +28,14 @@ export const useGridColumnPinningPreProcessors = (
 
   const prevAllPinnedColumns = React.useRef<string[]>([]);
 
+  const theme = useTheme();
+
   const reorderPinnedColumns = React.useCallback<GridPipeProcessor<'hydrateColumns'>>(
     (columnsState) => {
       eslintUseValue(pinnedColumns);
 
+      columnsState = updatePinnedColumns(columnsState as GridColumnsState, theme);
+
       if (columnsState.orderedFields.length === 0 || disableColumnPinning) {
         return columnsState;
       }
diff --git a/packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts b/packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts
index f5491e730..505e5e2a8 100644
--- a/packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts
@@ -410,8 +410,6 @@ export const createColumnsState = ({
     });
   }
 
-  columnsState = updatePinnedColumns(columnsState as GridColumnsState, theme);
-
   const columnsStateWithPreProcessing = apiRef.current.unstable_applyPipeProcessors(
     'hydrateColumns',
     columnsState,

```

The difference is that the `columnsState` in `hydrateColumns` includes the `__check__` column in the visible columns lookup.

Reply:
I found other underlying issues with pinning, I implemented a fix but I'm horrified by it :| Left a comment about `renderContext` & pinning.

Reply:
This has been fixed.